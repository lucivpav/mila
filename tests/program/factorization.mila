program factorization;

procedure factorization(n: integer);
var max, i: integer;
begin
    write('Factors of: ');
    writeln(n);

    if n < 4 then
    begin
        writeln(n);
        exit;
    end;

    while ((n mod 2) = 0) do
    begin
        writeln(2);
        n := n div 2;
    end;

    while ((n mod 3) = 0) do
    begin
        writeln(3);
        n := n div 3;
    end;

    max := n;
    i := 5;
    while i <= max do
    begin
        while ((n mod i) = 0) do
        begin
            writeln(i);
            n := n div i;
        end;
        i := i + 2;
        while ((n mod i) = 0) do
        begin
            writeln(i);
            n := n div i;
        end;
        i := i + 4;
    end;
    if n <> 1 then writeln(n);
end;

begin
    factorization(0);
    factorization(1);
    factorization(2);
    factorization(3);
    factorization(4);
    factorization(5);
    factorization(6);
    factorization(7);
    factorization(8);
    factorization(9);
    factorization(10);
    factorization(11);
    factorization(12);
    factorization(13);
    factorization(14);
    factorization(15);
    factorization(16);
    factorization(17);
    factorization(100);
    factorization(131);
    factorization(133);
end.
