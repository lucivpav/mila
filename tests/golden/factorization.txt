Factors of: 0
0
Factors of: 1
1
Factors of: 2
2
Factors of: 3
3
Factors of: 4
2
2
Factors of: 5
5
Factors of: 6
2
3
Factors of: 7
7
Factors of: 8
2
2
2
Factors of: 9
3
3
Factors of: 10
2
5
Factors of: 11
11
Factors of: 12
2
2
3
Factors of: 13
13
Factors of: 14
2
7
Factors of: 15
3
5
Factors of: 16
2
2
2
2
Factors of: 17
17
Factors of: 100
2
2
5
5
Factors of: 131
131
Factors of: 133
7
19
0
